src/flb_gzip_concatenated.c

```diff
@@ -0,0 +1,108 @@
+/* -*- Mode: C; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */
+
+/*  Fluent Bit
+ *  ==========
+ *  Copyright (C) 2015-2024 The Fluent Bit Authors
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#include <fluent-bit/flb_gzip_concatenated.h>
+
+static int vaild_os_flag(const char data)
+{
+    uint8_t p;
+
+    p = (uint8_t)data;
+    if (p == 0x00 || /* Fat Filesystem       */
+        p == 0x01 || /* Amiga                */
+        p == 0x02 || /* VMS                  */
+        p == 0x03 || /* Unix                 */
+        p == 0x04 || /* VM/CMS               */
+        p == 0x05 || /* Atari TOS            */
+        p == 0x06 || /* HPFS Filesystem (OS/2, NT) */
+        p == 0x07 || /* Macintosh            */
+        p == 0x08 || /* Z-System             */
+        p == 0x09 || /* CP/M                 */
+        p == 0x0a || /* TOPS-20              */
+        p == 0x0b || /* NTFS filesystem (NT) */
+        p == 0x0c || /* QDOS                 */
+        p == 0x0d || /* Acorn RISCOS         */
+        p == 0xff)   /* Unknown              */ {
+
+        return FLB_TRUE;
+    }
+
+    return FLB_FALSE;
+}
+
+size_t flb_gzip_concatenated_count(const char *data, size_t len)
+{
+    int i;
+    size_t count = 0;
+    const uint8_t *p;
+
+    p = (const uint8_t *) data;
+
+    /* search other gzip starting bits and method. */
+    for (i = 2; i < len &&
+                 i + 9 <= len; i++) {
+        /* A vaild gzip payloads are larger than 18 bytes. */
+        if (len - i < 18) {
+            break;
+        }
+
+        if (p[i] == 0x1F && p[i+1] == 0x8B && p[i+2] == 8 &&
+            vaild_os_flag(p[i+9])) {
+            count++;
+        }
+    }
+
+    return count;
+}
+
+size_t flb_gzip_concatenated_borders(const char *data, size_t len, size_t **out_borders, size_t border_count)
```

Comment:
since this function does pretty much the same than _count plus setting up the borders, I suggest to merge both and make **out_borders optional, e.g: if set, write the borders on it

Reply:
Thank you for your suggestion.
Fixed at https://github.com/fluent/fluent-bit/pull/9139/commits/dcc9ca51a74a28c095df95ac62fe3126452b4cf8.